internal/server/v1/propertyvalues/state.go

```diff
@@ -0,0 +1,184 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package propertyvalues
+
+import (
+	"container/heap"
+	"context"
+	"log"
+
+	cbt "cloud.google.com/go/bigtable"
+	pb "github.com/datacommonsorg/mixer/internal/proto"
+	"github.com/datacommonsorg/mixer/internal/store/bigtable"
+)
+
+type state struct {
+	property string
+	entity   string
+	limit    int
+	// CursorGroup that tracks the state of current data.
+	cursorGroup *pb.CursorGroup
+	// Raw entities read from BigTable
+	rawEntities [][]*pb.EntityInfo
+	// Total page count for each import group
+	// Merged entities
+	mergedEntities []*pb.EntityInfo
+	totalPage      map[int]int
+}
+
+type inState struct {
+	state
+	// Min heap for entity merge sort
+	heap *entityHeap
+}
+
+type outState struct {
+	state
+	// The used import group (only one import group is used for out property values).
+	usedImportGroup int
+	//
+}
+
+// init builds a new state given property, entity and cursor group.
+func (s *state) init(
+	ctx context.Context,
+	btGroup *bigtable.Group,
+	property string,
+	entity string,
+	limit int,
+	cursorGroup *pb.CursorGroup,
+	arcOut bool,
+) error {
+	// Constructor
+	s.property = property
+	s.entity = entity
+	s.cursorGroup = cursorGroup
+	s.rawEntities = nil
+	s.mergedEntities = nil
+	s.limit = limit
+	s.totalPage = map[int]int{}
+	// Read cache data based on pagination info.
+	rowListMap := map[int]cbt.RowList{}
+	for _, c := range cursorGroup.Cursors {
+		if c != nil {
+			rowList := buildSimpleRequestRowList(entity, property, arcOut, c.GetPage())
+			rowListMap[int(c.GetImportGroup())] = rowList
+		}
+	}
+	btDataList, err := bigtable.ReadWithGroupRowList(
+		ctx,
+		btGroup,
+		rowListMap,
+		action,
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+	// Store the raw cache data.
+	for idx, btData := range btDataList {
+		if data, ok := btData[entity]; ok {
+			pe := data.(*pb.PagedEntities)
+			s.rawEntities = append(s.rawEntities, pe.Entities)
+			s.totalPage[idx] = int(pe.TotalPageCount)
+		} else {
+			s.rawEntities = append(s.rawEntities, nil)
+		}
+	}
+	return nil
+}
+
+func (s *state) readNextPage(
+	ctx context.Context,
+	btGroup *bigtable.Group,
+	arcOut bool,
+	importGroup int,
+	page int32,
+) error {
+	log.Printf("Read new page: import group %d, page %d", importGroup, page)
+	rowList := buildSimpleRequestRowList(s.entity, s.property, arcOut, page)
+	rowListMap := map[int]cbt.RowList{
+		importGroup: rowList,
+	}
+	btDataList, err := bigtable.ReadWithGroupRowList(
+		ctx,
+		btGroup,
+		rowListMap,
+		action,
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+	if data, ok := btDataList[importGroup][s.entity]; ok {
```

Comment:
assert !arcOut || importGroup == s.usedImportGroup ?

Reply:
Here "s" is the base struct that does not have usedImportGroup field. The caller func nextOut() in property_values.go could guarantee this condition.